src/arch/riscv/pmp.cc

```diff
@@ -0,0 +1,223 @@
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution;
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "arch/riscv/pmp.hh"
+
+#include "arch/generic/tlb.hh"
+#include "arch/riscv/faults.hh"
+#include "arch/riscv/isa.hh"
+#include "base/addr_range.hh"
+#include "base/types.hh"
+#include "cpu/thread_context.hh"
+#include "math.h"
+#include "mem/request.hh"
+#include "params/PMP.hh"
+#include "sim/sim_object.hh"
+
+PMP::PMP(const Params &params) :
+    SimObject(params),
+    maxEntries(params.max_pmp),
+    numRules(0)
+{
+    for (int i=0; i < maxEntries; i++) {
+        PmpEntry entry;
+        pmpTable.emplace_back(entry);
+    }
+}
+
+Fault
+PMP::pmpCheck(const RequestPtr &req, BaseTLB::Mode mode,
+              RiscvISA::PrivilegeMode pmode, ThreadContext *tc)
+{
+    // First determine if pmp table should be consulted
+    if (!shouldCheckPMP(pmode, mode, tc))
+        return NoFault;
+
+    // An access should be successful if there are
+    // no rules defined yet or we are in M mode (based
+    // on specs v1.10)
+    if (numRules == 0 || (pmode == RiscvISA::PrivilegeMode::PRV_M))
+        return NoFault;
+
+    // all pmp entries need to be looked from the lowest to
+    // the highest number
+    int match_index = -1;
+    for (int i = 0; i < pmpTable.size(); i++) {
+        if (pmpTable[i].pmpAddr.contains(req->getPaddr())) {
+           //address matched
+            match_index = i;
+        }
+
+        if ((PMP_OFF != pmpGetAField(pmpTable[match_index].pmpCfg))
+                                            && (match_index > -1)) {
+            // check the RWX permissions from the pmp entry
+            uint8_t allowed_privs = PMP_READ | PMP_WRITE | PMP_EXEC;
+
+            // i is the index of pmp table which matched
+            allowed_privs &= pmpTable[match_index].pmpCfg;
+
+            // mode: R = 0, W = 1, X = 2
+            uint8_t mode_mask = 1 << mode;
```

Comment:
I'm not sure you can count on TLB::Mode to be 0, 1, or 2. Even though it increases the code, I think the full checks would be safer (e.g., `if (mode == TLB::Read)`). Let me know if I'm wrong.

Reply:
I think it can be a problem only if in the future the encoding of TLB::Mode is changed to something else. Otherwise, I can't think of a case where the use of mode_mask will be a problem. However, using mode == TLB::Read is definitely more readable and easy to understand. I can get rid of mode_mask if you suggest!

Reply:
Yeah, I'd do that